proposals/user-scripts-api.md

```diff
@@ -0,0 +1,163 @@
+  filter?: UserScriptFilter[],
+  callback?: function
+)
+browser.userScripts.update(
+  scripts?: RegisteredUserScript[],
+  callback?: function
+)
+```
+
+### Requirements
+
+#### A. A mechanism to execute code in the main world
+
+- User scripts can be registered in the `MAIN` world.
+- Extension can customize the `USER_SCRIPT` world CSP to inject a script tag into the host page (expanded in requirement B)
+
+#### B. The ability to execute code (with a separate CSP) in a world different from the main world and the extension's isolated world
+
+##### `USER_SCRIPT World
+
+- Isolated from the web page (similar to other isolated worlds), but will be un-permissioned.  It will not have access to any extension APIs or cross-origin exceptions (these don't exist in Chrome, but do in other browsers).
+- Exempt from the page's CSP (see Relax CSP).
+- Share DOM with the web page.  Code from both worlds cannot directly interact with each other, except through DOM APIs.
+- Can communicate with different JS worlds via `window.postMessage()`.  A dedicated API to communicate between worlds is being considered as part of future work.
+When an asymmetric security relationship may exist, the `MAIN` world is considered to be less privileged than the `USER_SCRIPT` world
+
+##### Relax CSP
+
+```
+browser.userScripts.configureWorld(
+  csp: string
+)
+```
+- Extension can customize the `USER_SCRIPT` world CSP. This will only affect scripts registered for the `USER_SCRIPT` world.  Scripts registered for the MAIN world will abide by the main world CSP.
+- By default, the `USER_SCRIPT` world CSP is undefined.
+- When `USER_SCRIPT` is undefined, the user agent will fall back to the extension’s `ISOLATED` world CSP
+
+#### C. A separate user script permission
+
+- A new extension permission will be added, and will be required in order to register user scripts.  This permission should be scoped specifically to the purpose of user scripts (as opposed to a general "remotely-hosted code" permission).
+- Browser vendors then can use it to:
+  - Inform their review pipeline that this is a particularly risky extension, and extra caution should be taken to ensure this is a valid use of the API.
+  - Reduce the likelihood of extensions using it for other purposes and allows browser vendors to act on any permissions that do as being non-compliant
+  - Present the user with different permission warnings, UI, or other gating, if they deem it necessary.
+
+#### D. Communication between JavaScript worlds
+
+As mentioned in requirement A, the user script world can communicate with different JS worlds via `window.postMessage()` and DOM APIs. New communication methods are being considered as potential future enhancements.
+
+###  Other considerations
+
+- When multiple scripts match and have the same runAt schedule, the execution order is:
+  - Scripts registered via the content_scripts key in the manifest file
+  - Scripts registered via scripting.registerContentScripts
```

Comment:
Where is `scripting.registerContentScripts` defined?

Reply:
https://developer.chrome.com/docs/extensions/reference/scripting/#method-registerContentScripts

Reply:
Please link that in the document.